content/en/blog/_posts/2021-03-09-Defining-NetworkPolicy-Conformance-For-CNIs

```diff
@@ -0,0 +1,373 @@
+It was clear that we needed to begin unifying the way we think about network policies in Kubernetes, if we wanted to avoid becoming an "ancient" API which was going to be overrun by implementation divergences. 
+
+For example:  
+- Calico as a CNI provider can be run using IPIP or VxLan mode.  CNI's such as Antrea and Cillium offer similar configuration divergences as well.
+- Some CNIs rely on IPTables for NetworkPolicies (i.e. calico), whereas others use a completely different technology stack (for example, the antrea project uses open vswith rules).
+- Some CNI providers only implement a subset of the Kubernete NetworkPolicy API (for example, certain providers don't support the ability to target a `NamedPort`, or a port range, which is a newer feature in the K8s API).
+
+Thus, end-users need to follow a multistep process to implement network policies to secure their applications:
+- Confirm that their CNI supports NetworkPolicys (some don't, such as flannel)
+- Confirm that their CNI supports the specific NetworkPolicy's that a user is interested in (again, the Named port or Port range examples come to mind here)
+- Confirm that their application's network policy definitions are doing the 'right thing'
+- Find out the vendor specific implementation of a policy, and check wether or not that implementation had a CNI neutral implementation (which is preferable for most users)
+
+The NetworkPolicy project in upstream Kubernetes aims at providing a community where people can learn about, and contribute to, the Kubernetes NetworkPolicy API and the surrounding ecosystem.
+
+## The First step: A validation framework for NetworkPolicies that was intuitive to use and understand
+
+The Kubernetes end to end suite has always had NetworkPolicy tests, but these weren't run in CI, and the way they were implemented didn't really provide any holistic, easily consumable information about how a policy was working in a cluster.
+This is because the original tests didn't provide any kind of visual summary of connectivity across a cluster.   We thus initially set out to make it easy to confirm CNI support for NetworkPolicies by
+making the end to end tests (which are often used by administrators or users to diagnose cluster conformance) easy to interpret.
+
+To solve the problem of confirming that CNI's support the basic features one cares about for a policy, we built a new NetworkPolicy validation tool into the Kubernete e2e framework which allows for visual inspection of policys and their effect on a standard set of pods in a cluster.
+For example, in the following test output.  As an example, we found a bug in [ovn kubernetes](https://github.com/ovn-org/ovn-kubernetes/issues/1782) with this tool that was really easy to characterize,
```

Comment:
I guess the beginning here (For example, in the following test output) can be removed, and keep only the sentence As an example, we've found a bug in...